Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h

```diff
@@ -0,0 +1,406 @@
+  GT gt = choose_param(get_param(snp, internal_np::geom_traits), GT());
+
+  CGAL_static_assertion((boost::is_same<Point, typename GT::Point_3>::value));
+
+  typedef typename GT::FT                                                             FT;
+  typedef CGAL::internal::Dynamic_property_map<vertex_descriptor, FT>       Default_tolerance_map;
+
+  typedef typename boost::lookup_named_param_def <
+    internal_np::tolerance_map_t,
+    SourceNamedParameters,
+    Default_tolerance_map // default
+  > ::type                                                                  Tolerance_map;
+
+  Tolerance_map tol_vm = choose_param(get_param(snp, internal_np::tolerance_map),
+                                      Default_tolerance_map());
+
+  // Extract border vertices
+  std::vector<halfedge_descriptor> border_vertices;
+  BOOST_FOREACH(edge_descriptor ed, edges(smesh))
+  {
+    if(is_border(ed, smesh))
+    {
+      halfedge_descriptor hd = halfedge(ed, smesh);
+      if(!is_border(hd, smesh))
+        hd = opposite(hd, smesh);
+      border_vertices.push_back(hd);
+    }
+  }
+
+  // If no tolerance is provided, compute a unique bound at each vertex
+  if(boost::is_same<Tolerance_map, Default_tolerance_map>::value)
+    compute_tolerance_at_vertices(tol_vm, border_vertices, svpm, gt, smesh);
+
+  // Try to snap vertices
+  std::vector<Box> boxes;
+  BOOST_FOREACH(halfedge_descriptor hd, border_vertices)
+  {
+    vertex_descriptor vd = target(hd, smesh);
+    const FT eps = get(tol_vm, vd);
+
+    const typename GT::Vector_3 eps_v = gt.construct_vector_3_object()(eps, eps, eps);
+    const typename GT::Vector_3 opp_eps_v = gt.construct_vector_3_object()(-eps, -eps, -eps);
+    const Point& pm = gt.construct_translated_point_3_object()(get(svpm, vd), opp_eps_v);
+    const Point& pM = gt.construct_translated_point_3_object()(get(svpm, vd), eps_v);
+
+    Bbox_3 bbox = gt.construct_bbox_3_object()(pm) + gt.construct_bbox_3_object()(pM);
+    boxes.push_back(Box(bbox, vd));
+  }
+
+  std::vector<Box> target_boxes;
+  BOOST_FOREACH(vertex_descriptor vd, vertices(tmesh))
+  {
+    const Point& p = get(tvpm, vd);
+    target_boxes.push_back(Box(gt.construct_bbox_3_object()(p), vd));
+  }
+
+  // the correspondence map
+  boost::unordered_map<vertex_descriptor/*source*/, vertex_descriptor/*target*/> vertex_map;
+
+  // Shenanigans to pass a reference as callback (which is copied by value by 'box_intersection_d')
+  typedef Vertex_proximity_report<PolygonMesh, GT, SVPM, TVPM, Tolerance_map, Box>     Reporter;
+  Reporter vpr(vertex_map, svpm, tvpm, tol_vm, gt);
+  boost::function<void(const Box&, const Box&)> callback(boost::ref(vpr));
+
+  CGAL::box_intersection_d(boxes.begin(), boxes.end(),
+                           target_boxes.begin(), target_boxes.end(),
+                           callback);
+
+  if(!vpr.is_valid_snapping())
+  {
+    std::cerr << "Aborting snapping operation. Choose a better value for epsilon!" << std::endl;
```

Comment:
IIRC, we can still do the snapping for those without conflicts. A return value could indicate if every snap was correctly done. What do you think?